network-area-diagram/src/main/java/com/powsybl/nad/build/iidm/VoltageLevelFilter.java

```diff
@@ -85,18 +91,71 @@ public static Collection<VoltageLevel> getNextDepthVoltageLevels(Network network
         return voltageLevelSet;
     }
 
-    private static void traverseVoltageLevels(Set<VoltageLevel> voltageLevelsDepth, int depth, Set<VoltageLevel> visitedVoltageLevels) {
+    private static void traverseVoltageLevels(Set<VoltageLevel> voltageLevelsDepth, int depth, Set<VoltageLevel> visitedVoltageLevels, Predicate<VoltageLevel> predicate) {
         if (depth < 0) {
             return;
         }
         Set<VoltageLevel> nextDepthVoltageLevels = new HashSet<>();
         for (VoltageLevel vl : voltageLevelsDepth) {
-            if (!visitedVoltageLevels.contains(vl)) {
+            if (!visitedVoltageLevels.contains(vl) && predicate.test(vl)) {
                 visitedVoltageLevels.add(vl);
                 vl.visitEquipments(new VlVisitor(nextDepthVoltageLevels, visitedVoltageLevels));
             }
         }
-        traverseVoltageLevels(nextDepthVoltageLevels, depth - 1, visitedVoltageLevels);
+        traverseVoltageLevels(nextDepthVoltageLevels, depth - 1, visitedVoltageLevels, predicate);
+    }
+
+    private static void traverseVoltageLevels(Set<VoltageLevel> voltageLevelsDepth, int depth, Set<VoltageLevel> visitedVoltageLevels) {
+        traverseVoltageLevels(voltageLevelsDepth, depth, visitedVoltageLevels, NO_FILTER);
+    }
+
+    private static void traverseVoltageLevelsWithVoltageFilter(Set<VoltageLevel> voltageLevelsDepth, int depth, Set<VoltageLevel> visitedVoltageLevels,
+                                                               double lowNominalVoltageBound, double highNominalVoltageBound) {
+
+        Predicate<VoltageLevel> voltageFilterPredicate = voltageLevel -> voltageLevel.getNominalV() >= lowNominalVoltageBound && voltageLevel.getNominalV() <= highNominalVoltageBound;
+        traverseVoltageLevels(voltageLevelsDepth, depth, visitedVoltageLevels, voltageFilterPredicate);
+    }
+
+    public static VoltageLevelFilter createNominalVoltageFilter(Network network, List<String> voltageLevelIds,
+                                                                double lowNominalVoltageBound, double highNominalVoltageBound,
+                                                                int depth) {
+        Objects.requireNonNull(network);
+        Objects.requireNonNull(voltageLevelIds);
+        checkVoltageBoundValues(lowNominalVoltageBound, highNominalVoltageBound);
+        Set<VoltageLevel> startingSet = new HashSet<>();
+
+        for (String voltageLevelId : voltageLevelIds) {
+            VoltageLevel vl = network.getVoltageLevel(voltageLevelId);
+            if (vl == null) {
+                throw new PowsyblException(UNKNOWN_VOLTAGE_LEVEL + voltageLevelId + "'");
+            }
+            if (vl.getNominalV() < lowNominalVoltageBound || vl.getNominalV() > highNominalVoltageBound) {
```

Comment:
you could also use the predicate here, and in fact here it's the same code as in `createVoltageLevelsDepthFilter` so you could factorize the whole method (except the `checkBounds` which is specific to that filter)